src/v/cloud_topics/dl_stm/dl_stm_state.cc

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cloud_topics/dl_stm/dl_stm_state.h"
+
+#include "model/fundamental.h"
+
+namespace experimental::cloud_topics {
+
+bool dl_stm_state::register_overlay_cmd(const dl_overlay& o) noexcept {
+    // Validate command
+    if (o.base_offset == kafka::offset{}) {
+        return false;
+    }
+    if (o.base_offset > o.last_offset) {
+        return false;
+    }
+    if (o.size_bytes() == 0) {
+        return false;
+    }
+    // Admit command
+    _overlays.append(o);
+
+    // Propagate last reconciled offset
+    propagate_last_reconciled_offset(o.base_offset, o.last_offset);
+
+    // Update terms
+    for (auto kv : o.terms) {
+        _term_map.insert(kv);
+    }
+    return true;
+}
+
+std::optional<dl_overlay>
+dl_stm_state::lower_bound(kafka::offset o) const noexcept {
+    return _overlays.lower_bound(o);
+}
+
+std::optional<kafka::offset>
+dl_stm_state::get_term_last_offset(model::term_id t) const noexcept {
+    auto it = _term_map.upper_bound(t);
+    if (it == _term_map.end()) {
+        return std::nullopt;
+    }
+    return kafka::prev_offset(it->second);
+}
+
+void dl_stm_state::propagate_last_reconciled_offset(
+  kafka::offset base, kafka::offset last) noexcept {
+    // If newly added offset range connects with the last reconciled offset
```

Comment:
this method would benefit being structured as a series of `else if` blocks so we have a list of explicit cases that we're handling and an `else` at the end with an assert to catch flaws in our logic and having to blindly trust comments like `// Invariant: here prev_offset(base) <= LRO.`

---

it can also benefit being extracted as a `struct last_reconciled_offset_tracker` so that it is very simple to test in isolation. i've extracted similar-ish logic in batch cache with [dirty tracker](https://github.com/redpanda-data/redpanda/blob/dc2b959910a476ea3a60529e6ac1328e21705334/src/v/storage/batch_cache.h#L472)